[Client] 회원가입 mock API를 작성하고 회원가입 페이지에 연동한다

leegwae opened the pull request:
## 🤷‍♂️ Description
- #20 

<!-- 구현하고자 하는 기능에 대해 작성해 주세요. -->

## 📝 Primary Commits

<!-- 세부 구현 사항을 리스트로 작성해주세요. -->

- [X] `POST /api/auth/signup` mock API 작성
- [X] `POST /api/auth/signup` API 요청 함수 작성
- [X] `POST /api/auth/signup` API 뮤테이션 훅 작성
- [X] 회원가입 페이지 `회원가입` 버튼에 `POST /api/auth/signup` API 연동

leegwae committed:
design: 에러 메시지의 색을 빨강으로 바꾼다

leegwae committed:
feat: `forwardRef`로 감싼 컴포넌트에 ref를 전달한다

leegwae committed:
feat: `POST /api/user/auth` 엔드포인트를 상수로 정의한다 (#20)

leegwae committed:
feat: `POST /api/auth/signup` API 요청 함수를 정의한다 (#20)

leegwae committed:
feat: `useSignUpMutation`을 정의한다 (#20)

leegwae committed:
feat: `/api/auth/signup` mock API를 정의한다 (#20)

leegwae committed:
feat: 회원가입 페이지에 `POST /api/auth/signup` API를 연동한다 (#20)

leegwae commented:
https://github.com/blacktokkies/toquiz/blob/d640244eabcc8b93d4bc7350c52c39c28c1b8850/client/src/lib/api/auth.ts#L18-L22

1. 위 인터페이스는 shared 패키지로 옮겨야함.
2. 현재 validation은 적용 안되어있음

soomanbaek commented:
> https://github.com/blacktokkies/toquiz/blob/d640244eabcc8b93d4bc7350c52c39c28c1b8850/client/src/lib/api/auth.ts#L18-L22
> 
> 1. 위 인터페이스는 shared 패키지로 옮겨야함.
> 2. 현재 validation은 적용 안되어있음

shared 패키지로 옮겼습니다~

soomanbaek commented on client/src/mocks/handlers/auth.ts:
```diff
@@ -0,0 +1,22 @@
+
+import { apiUrl } from '@/lib/constants';
+
+export const signUp = rest.post<SignUpParams, never, SignUpResponse>(
```
SingUpParams를 shared폴더에 SignUpBody라는 이름의 인터페이스로 옮겨 놓아서 `shared/src/api/users/auth/signup.ts`의 SignUpBody로 받아오시면 될거 같아요

soomanbaek commented on client/src/mocks/handlers/auth.ts:
```diff
@@ -0,0 +1,22 @@
+  async (req, res, ctx) => {
+    const { nickname }: SignUpParams = await req.json();
+
+    return res(
```
response Mocking하신 부분인거 같은데, shared 폴더에 signUpResponse를 활용하셔도 좋을거 같아요!

soomanbaek commented on client/src/lib/constants.ts:
```diff
@@ -0,0 +1,15 @@
+};
+
+const authApiUrl = {
+  signup: () => `/api/user/auth`,
```
API URI 수정에 따라 signup api 요청을 `/api/users/auth/signup` 으로 보내시면 돼요!

soomanbaek commented on client/src/lib/api/auth.ts:
```diff
@@ -0,0 +1,22 @@
+import { apiUrl } from '@/lib/constants';
+
+// TODO: 정의한 fetch 모듈로 바꿔야한다
+export const signUp = async (params: SignUpParams): Promise<any> =>
```
반환 타입이 `Promise<any>` 로 하신 이유가 있을까요?? 아니면 아직 TODO가 있어 임시로 any를 사용하신건가요??

leegwae replied:
아직 todo가 있어 임시로 any를 사용했습니다^^

soomanbaek commented on client/src/lib/api/auth.ts:
```diff
@@ -0,0 +1,22 @@
+    body: JSON.stringify(params),
+  }).then(async (res) => res.json());
+
+export interface User {
+  username: string;
+  nickname: string;
+}
+
+export interface SignUpParams {
+  username: User['username'];
+  password: string;
+  nickname: User['nickname'];
+}
```
해당 부분은 shared 폴더에 추가했습니다

soomanbaek reviewed (approved):
수고하셨어요!

추가적으로, SignUp API가 `api/auth/signup`에서 `api/users/auth/signup`으로 바뀌어서 그 부분 수정해주시면 될거 같아요!

leegwae committed:
fix: 모듈 진입점으로 `index.ts`를 만든다

- `package.json`에 `main`을 `index.js`로 명시했으므로 해당 빌드물로 산출될 `index.ts`를 만들어준다

leegwae committed:
fix: Indexed Access Type을 제거한다

- 현재 빌드된 Indexed Access Type이 제대로 된 타입을 가져오지 못하고 있음 (`User['nickname']`으로 명시한 `SignUpBody`의 `nickname`이 `User`로 가져와지고 있음)

leegwae committed:
feat: POST 회원가입 API 엔드포인트 상수를 변경한다 (#20)

leegwae committed:
refactor: Shared 패키지로부터 API 인터페이스를 가져다쓴다 (#20)

leegwae commented:
### shared의 SignUpBody 인터페이스에서 Indexed Access Type을 제거하였습니다.

이슈는 #25 를 참고하세요.

soomanbaek commented on client/src/lib/constants.ts:
```diff
@@ -0,0 +1,15 @@
+};
+
+const authApiUrl = {
+  signup: () => `/api/user/auth/signup`,
```
회원가입 url은 `/api/users/auth/signup` 이에요 (`user` -> `users`)

soomanbaek reviewed (approved):
수고하셨어요~! 
추가적으로 shared 폴더 수정은 client, server 둘 다 영향이 가서 브렌치를 따로 파서 수정하는게 좋아 보여요!

leegwae committed:
fix: POST 회원가입 API 엔드포인트를 수정한다 (#20)
